pyccel/ast/internals.py

```diff
@@ -169,3 +168,69 @@ def __str__(self):
             stop = str(self.stop)
         return '{0} : {1}'.format(start, stop)
 
+class Symbol(Basic):
+    """Symbolic placeholder for a Python variable, which has a name but no type yet.
+    This is very generic, and it can also represent a function or a module.
+
+    Parameters
+    ----------
+    name : String
+        name of the symbol
+
+    Examples
+    --------
+    >>> from pyccel.ast.internals import Symbol
+    >>> x = Symbol('x')
+    x
+    """
+    def __init__(self, name):
+        if not isinstance(name, str):
+            raise TypeError('Symbol name should be a string, not '+ str(type(name)))
+        self._name = name
+        super().__init__()
+
+    @property
+    def name(self):
+        """Name of the symbol
+        """
+        return self._name
+
+    def __eq__(self, other):
+        return self._name == other
```

Comment:
Shouldn't you check the type before? E.g.
```python
def __eq__(self, other):

    if type(self) is type(other):
        return self._name == other._name

    return False
```
Hashes don't have to be picky, but equality checks must be!

Reply:
The `__eq__` operator of `Symbol` is designed such that:
```
Symbol('s')==str('s')
```
returns `True` to avoid the hundreds of casts done all over pyccel. It is just a wrapper around the string

Reply:
Not sure I understand why mimicking Sympy is necessary. Can you give me an example?

Reply:
```
a = 'name'
sym = Symbol(''name)
```
to avoid writing ```str(sym) == a``` every time were not sure if it's a symbol or string
we can do ```a == sym```  just as Emily said Symbol is just a wrapper around the string

Reply:
I see. But then why do we even need a Symbol class? We could just use strings and be happy, no?

Reply:
If we care about performance, we should be sure of the types that we use at all times. Otherwise we should call `str` and be satisfied with the performance loss introduced by the `__str__` method...

Reply:
> I see. But then why do we even need a Symbol class? We could just use strings and be happy, no?

As I understand it, the wrapper is used to choose the correct `_visit_X` method in the `SemanticParser`. A `Symbol` is something that represents an object (variable/function/datatype/etc), while a string represents a string.

Reply:
That said, I have just checked [Codacy](https://app.codacy.com/gh/pyccel/pyccel/file/52912287204/coverage?bid=15415567&fileBranchId=15415567) and it seems the `_visit_str` function is not used. Presumably it has been replaced with `_visit_LiteralString` so  perhaps we don't need a symbol class anymore

Reply:
> > I see. But then why do we even need a Symbol class? We could just use strings and be happy, no?
> 
> As I understand it, the wrapper is used to choose the correct `_visit_X` method in the `SemanticParser`. A `Symbol` is something that represents an object (variable/function/datatype/etc), while a string represents a string.

I was being ironical, but you see.. we could still achieve what you describe with a dummy class that subclasses `str`. Python strings are immutable, and hence hashable by definition...

Reply:
Hmm, that sounds like a good idea to me, although we would have to track down all the uses of `.name`

Reply:
a Symbol class is like a `Name` class, it just holds the name, it should be kept though

Reply:
> a Symbol class is like a `Name` class, it just holds the name, it should be kept though,
> the Python AST uses the class `Name`

@saidctb Emily and I were discussing about using this extremely simple class:
```python
class Symbol(str):
    pass

x = Symbol('hello')
```
Here the comparison `x == 'hello'` will return `True`.

Reply:
ah okay , yes maybe

Reply:
I think the class should inherit from Basic though so it can have an `fst` which lets us say at which line/column there is a problem

Reply:
OK, let's leave things as they are for now. :smiley:

Reply:
> I think the class should inherit from Basic though so it can have an `fst` which lets us say at which line/column there is a problem

I am not so sure about this anymore. Currently we use sympy's `Symbol`, which does not have the `fst` attribute. I imagine that in 99.9 % of cases (if not all of them) the problem is generated at another node that uses one or more `Symbols` objects, so we would still have access to the `fst` through that node.

Reply:
This is one of the things that is improved in PR #718. As almost everything that is returned from `_visit` will then inherit from Basic, the fst can be set more consistently. In this case it is then easier to extract the fst from the `symbol` argument passed to `errors.report` rather than passing a `bounding_box`. Although the problem is not generated by the `Symbol` node, the `Symbol` node is used as the `symbol` passed to `errors.report` I think